## Project Title


Revolut UK Transaction Performance & Business Intelligence Report

## Background

Revolut is a UK-based fintech company that provides digital banking services, including payments, currency exchange, money transfers, and financial management solutions across multiple subscription plans.
As transaction volumes continue to grow, understanding customer behavior, revenue performance, fraud patterns, and operational efficiency becomes critical for driving business growth, improving customer experience, and minimizing financial risk.


## Business Problem

Revolut’s analytics team has identified gaps in understanding the key drivers of revenue growth, areas where fraudulent activities are most concentrated, the most valuable customer segments, and operational patterns affecting transaction success rates.
Without clear visibility into these areas, strategic decisions related to product development, fraud prevention, customer retention, and operational optimization remain reactive rather than fully data-driven.


## Project Aim & Objectives

To analyse Revolut’s transaction data and deliver actionable business intelligence that supports strategic decision-making across revenue, fraud, customer behavior, and operational performance.
1. Analyse transaction trends over time to identify peak periods and seasonal patterns
2. Evaluate revenue performance across subscription plans, transaction types and categories
3. Identify fraud patterns by time, country, category and customer segment
4. Understand customer behavior across age groups, cities and payment methods
5. Assess transaction success and failure rates across channels
6. Provide strategic recommendations to improve revenue, reduce fraud and enhance customer experience.



## Import Libraries

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

## Load Dataset

In [2]:
df= pd.read_csv(r"C:\Users\computer1\Desktop\Revolut-Analysis\Data\revolut_dataset.csv")
df

,transaction_id,user_id,transaction_date,transaction_time,transaction_type,transaction_category,amount,currency,transaction_status,payment_method,age_group,country,city,subscription_plan,fraud_flag,transaction_fee,revenue_generated,cashback_amount
0,TXN007094,USR93203,1/1/2023,9:12:25,Transfer,Utilities,850.38,EUR,Successful,Card,45-54,France,Bordeaux,Plus,0,8.50,6.80,4.25
1,TXN011155,USR17160,1/1/2023,13:45:55,Transfer,Shopping,4745.29,GBP,Successful,Bank Transfer,25-34,United Kingdom,Glasgow,Standard,1,71.18,56.94,0.00
2,TXN002651,USR24022,1/1/2023,15:11:31,Purchase,NaN,409.40,GBP,Successful,Apple Pay,25-34,United Kingdom,Manchester,Standard,0,6.14,4.91,0.00
3,TXN011642,USR85377,1/1/2023,21:01:06,Purchase,Food & Dining,643.73,EUR,Successful,Google Pay,35-44,France,Marseille,Metal,0,1.29,1.03,9.66
4,TXN003802,USR80439,1/1/2023,22:24:18,Purchase,Shopping,143.16,GBP,Successful,Apple Pay,25-34,United Kingdom,Bristol,Metal,0,0.29,0.23,2.15
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14995,TXN001957,USR55745,12/31/2024,8:55:37,Top-up,Shopping,2021.10,GBP,Failed,Card,45-54,United Kingdom,Leeds,Plus,0,20.21,16.17,NaN
14996,TXN007274,USR85454,12/31/2024,20:56:37,Purchase,Healthcare,418.49,EUR,Successful,Bank Transfer,18-24,Germany,Berlin,Standard,0,6.28,5.02,0.00
14997,TXN009711,USR55398,12/31/2024,22:09:14,Purchase,Shopping,630.43,GBP,Successful,Card,45-54,United Kingdom,London,Standard,0,9.46,7.57,0.00
14998,TXN014078,USR60014,12/31/2024,12:36:09,Transfer,Food & Dining,2544.17,USD,Successful,Card,25-34,United States,Chicago,Plus,0,25.44,20.35,12.72


## Understanding the Data

In [3]:
df.head()

,transaction_id,user_id,transaction_date,transaction_time,transaction_type,transaction_category,amount,currency,transaction_status,payment_method,age_group,country,city,subscription_plan,fraud_flag,transaction_fee,revenue_generated,cashback_amount
0,TXN007094,USR93203,1/1/2023,9:12:25,Transfer,Utilities,850.38,EUR,Successful,Card,45-54,France,Bordeaux,Plus,0,8.50,6.80,4.25
1,TXN011155,USR17160,1/1/2023,13:45:55,Transfer,Shopping,4745.29,GBP,Successful,Bank Transfer,25-34,United Kingdom,Glasgow,Standard,1,71.18,56.94,0.00
2,TXN002651,USR24022,1/1/2023,15:11:31,Purchase,NaN,409.40,GBP,Successful,Apple Pay,25-34,United Kingdom,Manchester,Standard,0,6.14,4.91,0.00
3,TXN011642,USR85377,1/1/2023,21:01:06,Purchase,Food & Dining,643.73,EUR,Successful,Google Pay,35-44,France,Marseille,Metal,0,1.29,1.03,9.66
4,TXN003802,USR80439,1/1/2023,22:24:18,Purchase,Shopping,143.16,GBP,Successful,Apple Pay,25-34,United Kingdom,Bristol,Metal,0,0.29,0.23,2.15


In [4]:
df.tail()

,transaction_id,user_id,transaction_date,transaction_time,transaction_type,transaction_category,amount,currency,transaction_status,payment_method,age_group,country,city,subscription_plan,fraud_flag,transaction_fee,revenue_generated,cashback_amount
14995,TXN001957,USR55745,12/31/2024,8:55:37,Top-up,Shopping,2021.10,GBP,Failed,Card,45-54,United Kingdom,Leeds,Plus,0,20.21,16.17,NaN
14996,TXN007274,USR85454,12/31/2024,20:56:37,Purchase,Healthcare,418.49,EUR,Successful,Bank Transfer,18-24,Germany,Berlin,Standard,0,6.28,5.02,0.00
14997,TXN009711,USR55398,12/31/2024,22:09:14,Purchase,Shopping,630.43,GBP,Successful,Card,45-54,United Kingdom,London,Standard,0,9.46,7.57,0.00
14998,TXN014078,USR60014,12/31/2024,12:36:09,Transfer,Food & Dining,2544.17,USD,Successful,Card,25-34,United States,Chicago,Plus,0,25.44,20.35,12.72
14999,TXN001711,USR46467,12/31/2024,19:47:01,Transfer,Food & Dining,1283.99,USD,Successful,Card,35-44,United States,New York,Standard,0,19.26,15.41,0.00


#### Observations
- Dataset spans January 2023 to December 2024 confirming the full 2-year range
- Missing values are visible in transaction_category and cashback_amount columns
- Currency values correctly reflect the customer's country
- Cashback is 0.00 for Standard plan users as expected

In [5]:
df.shape

(15000, 18)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   transaction_id        15000 non-null  object 
 1   user_id               15000 non-null  object 
 2   transaction_date      15000 non-null  object 
 3   transaction_time      15000 non-null  object 
 4   transaction_type      15000 non-null  object 
 5   transaction_category  14775 non-null  object 
 6   amount                15000 non-null  float64
 7   currency              15000 non-null  object 
 8   transaction_status    15000 non-null  object 
 9   payment_method        14850 non-null  object 
 10  age_group             14775 non-null  object 
 11  country               15000 non-null  object 
 12  city                  14700 non-null  object 
 13  subscription_plan     15000 non-null  object 
 14  fraud_flag            15000 non-null  int64  
 15  transaction_fee    

#### Observations
15,000 rows and 18 columns confirmed

5 columns contain missing values:
 - transaction_category: 225 missing
 - payment_method: 150 missing
 - age_group: 225 missing
 - city: 300 missing
- cashback_amount: 668 missing
- transaction_date and transaction_time are stored as 
 object (string) — need converting to datetime
- All other data types look appropriate

In [7]:
df.describe()

,amount,fraud_flag,transaction_fee,revenue_generated,cashback_amount
count,15000.000000,15000.000000,15000.000000,15000.000000,14332.000000
mean,1089.679413,0.039133,10.971601,8.777256,5.049705
std,1364.862718,0.193919,14.179018,11.343210,13.573032
min,2.050000,0.000000,0.010000,0.010000,0.000000
25%,309.350000,0.000000,2.770000,2.220000,0.000000
50%,607.145000,0.000000,5.980000,4.780000,0.000000
75%,1175.617500,0.000000,11.240000,8.990000,4.060000
max,9965.140000,1.000000,74.980000,59.980000,149.480000


#### Observations
- Average transaction amount is £1,089.68 with a wide 
  range from £2.05 to £9,965.14
- Only 4% of transactions are flagged as fraud — realistic
- 50% of cashback values are 0.00 confirming Standard 
  plan users receive no cashback
- High standard deviation on amount (£1,364) confirms 
  significant variation in transaction sizes

## Checking for duplicates

In [8]:
print(f"Duplicate rows: {df.duplicated().sum()}")


Duplicate rows: 0


In [9]:
print(f"Duplicate transaction IDs: {df['transaction_id'].duplicated().sum()}")

Duplicate transaction IDs: 0


## Checking for unique values in categorical columns 

In [10]:
cat_cols = [
 'transaction_type', 'transaction_category', 
 'transaction_status', 'payment_method', 
 'subscription_plan', 'currency']

for col in cat_cols:
 print(f"{col}: {df[col].unique()}")

transaction_type: ['Transfer' 'Purchase' 'Top-up' 'Refund' 'Withdrawal']
transaction_category: ['Utilities' 'Shopping' nan 'Food & Dining' 'Transfer' 'Healthcare'
 'Other' 'Entertainment' 'Travel']
transaction_status: ['Successful' 'Failed' 'Reversed']
payment_method: ['Card' 'Bank Transfer' 'Apple Pay' 'Google Pay' 'Direct Debit' nan]
subscription_plan: ['Plus' 'Standard' 'Metal' 'Premium']
currency: ['EUR' 'GBP' 'USD']


#### Observations
- No duplicate rows or transaction IDs found — 
  dataset integrity is confirmed
- Categorical columns are consistent with no 
  typos or casing issues
- NaN values confirmed in transaction_category 
  and payment_method as expected
- All categories align with expected Revolut 
  transaction types and payment methods

## Data Cleaning
In this section we address the issues identified during 
data understanding, fixing data types and handling 
missing values

#### convert transaction_ date to datetime

In [11]:
df['transaction_date'] = pd.to_datetime(df['transaction_date'])
print(df['transaction_date'].dtype)

datetime64[ns]


In [12]:
df.head()

,transaction_id,user_id,transaction_date,transaction_time,transaction_type,transaction_category,amount,currency,transaction_status,payment_method,age_group,country,city,subscription_plan,fraud_flag,transaction_fee,revenue_generated,cashback_amount
0,TXN007094,USR93203,2023-01-01,9:12:25,Transfer,Utilities,850.38,EUR,Successful,Card,45-54,France,Bordeaux,Plus,0,8.50,6.80,4.25
1,TXN011155,USR17160,2023-01-01,13:45:55,Transfer,Shopping,4745.29,GBP,Successful,Bank Transfer,25-34,United Kingdom,Glasgow,Standard,1,71.18,56.94,0.00
2,TXN002651,USR24022,2023-01-01,15:11:31,Purchase,NaN,409.40,GBP,Successful,Apple Pay,25-34,United Kingdom,Manchester,Standard,0,6.14,4.91,0.00
3,TXN011642,USR85377,2023-01-01,21:01:06,Purchase,Food & Dining,643.73,EUR,Successful,Google Pay,35-44,France,Marseille,Metal,0,1.29,1.03,9.66
4,TXN003802,USR80439,2023-01-01,22:24:18,Purchase,Shopping,143.16,GBP,Successful,Apple Pay,25-34,United Kingdom,Bristol,Metal,0,0.29,0.23,2.15


#### Handling missing values 

In [13]:
print("Missing values before cleaning:")
print(df.isnull().sum()[df.isnull().sum() > 0])

Missing values before cleaning:
transaction_category    225
payment_method          150
age_group               225
city                    300
cashback_amount         668
dtype: int64


##### Replace transaction_category with Uncategorised 
**Reasoning:** The transaction_category column contained 
225 missing values (1.5% of the dataset). Since these 
transactions exist but lack a category label, they were 
filled with 'Uncategorised' rather than being dropped or 
filled with an assumed category. This approach preserves 
all 15,000 records while maintaining transparency about 
the data gap. The small percentage means this decision 
has minimal impact on the overall category analysis.

In [14]:
df["transaction_category"] = df["transaction_category"].fillna("Uncategorised")
print(f"transaction_category missing: {df['transaction_category'].isnull().sum()}")
df


transaction_category missing: 0


,transaction_id,user_id,transaction_date,transaction_time,transaction_type,transaction_category,amount,currency,transaction_status,payment_method,age_group,country,city,subscription_plan,fraud_flag,transaction_fee,revenue_generated,cashback_amount
0,TXN007094,USR93203,2023-01-01,9:12:25,Transfer,Utilities,850.38,EUR,Successful,Card,45-54,France,Bordeaux,Plus,0,8.50,6.80,4.25
1,TXN011155,USR17160,2023-01-01,13:45:55,Transfer,Shopping,4745.29,GBP,Successful,Bank Transfer,25-34,United Kingdom,Glasgow,Standard,1,71.18,56.94,0.00
2,TXN002651,USR24022,2023-01-01,15:11:31,Purchase,Uncategorised,409.40,GBP,Successful,Apple Pay,25-34,United Kingdom,Manchester,Standard,0,6.14,4.91,0.00
3,TXN011642,USR85377,2023-01-01,21:01:06,Purchase,Food & Dining,643.73,EUR,Successful,Google Pay,35-44,France,Marseille,Metal,0,1.29,1.03,9.66
4,TXN003802,USR80439,2023-01-01,22:24:18,Purchase,Shopping,143.16,GBP,Successful,Apple Pay,25-34,United Kingdom,Bristol,Metal,0,0.29,0.23,2.15
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14995,TXN001957,USR55745,2024-12-31,8:55:37,Top-up,Shopping,2021.10,GBP,Failed,Card,45-54,United Kingdom,Leeds,Plus,0,20.21,16.17,NaN
14996,TXN007274,USR85454,2024-12-31,20:56:37,Purchase,Healthcare,418.49,EUR,Successful,Bank Transfer,18-24,Germany,Berlin,Standard,0,6.28,5.02,0.00
14997,TXN009711,USR55398,2024-12-31,22:09:14,Purchase,Shopping,630.43,GBP,Successful,Card,45-54,United Kingdom,London,Standard,0,9.46,7.57,0.00
14998,TXN014078,USR60014,2024-12-31,12:36:09,Transfer,Food & Dining,2544.17,USD,Successful,Card,25-34,United States,Chicago,Plus,0,25.44,20.35,12.72


##### Fill payment method with mode
**Reasoning:** The payment_method column contained 150 
missing values (1.0% of the dataset). Since payment method 
is a categorical column with no numerical meaning, it was 
filled with the mode (most frequently occurring value). 
This is a standard approach for categorical missing values 
where the missingness appears random rather than systematic. 
The small percentage ensures minimal impact on payment 
method analysis.

In [15]:
mode_payment = df['payment_method'].mode()[0]
df['payment_method'] = df['payment_method'].fillna(mode_payment)
print(f"payment_method missing: {df['payment_method'].isnull().sum()}")
print(f"Filled with: {mode_payment}")


payment_method missing: 0
Filled with: Card


##### Fill age group with Unknown
**Reasoning:** The age_group column contained 225 missing values 
representing users who did not complete their profile 
during signup. Since age group cannot be inferred from 
other columns, these were filled with 'Unknown' to 
preserve all records while clearly identifying 
incomplete profiles. This affects only 1.5% of the 
dataset and has minimal impact on demographic analysis.

In [16]:
df['age_group'] = df['age_group'].fillna('Unknown')

print(f"age_group missing: {df['age_group'].isnull().sum()}")


age_group missing: 0


#### Fill city with Unknown
**Reasoning:**  The city column contained 300 missing values where 
users did not share their location. Since city cannot 
be reliably inferred from other available columns, 
these were filled with 'Unknown'. This represents 2% 
of the dataset and has minimal impact on geographic 
analysis.

In [17]:
df['city'] = df['city'].fillna('Unknown')

print(f"city missing: {df['city'].isnull().sum()}")

city missing: 0


#### Fill cash back amount with 0
**Reasoning:** The cashback_amount column contained 668 missing values 
corresponding to Failed and Reversed transactions. Since 
cashback is only awarded on successful transactions under 
Revolut's subscription model, these missing values were 
filled with 0. This is a business logic decision rather 
than a statistical assumption, failed transactions 
genuinely generate no cashback regardless of subscription 
plan.

In [18]:
df['cashback_amount'] = df['cashback_amount'].fillna(0)

print(f"cashback_amount missing: {df['cashback_amount'].isnull().sum()}")

cashback_amount missing: 0


#### Outlier detection using IQR method (Percentile)

In [19]:
numerical_cols = ['amount', 'transaction_fee', 'revenue_generated']

for col in numerical_cols:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    outliers = df[(df[col] < lower) | (df[col] > upper)]
    
    print(f"{col}:")
    print(f"  Q1: {Q1:.2f} | Q3: {Q3:.2f} | IQR: {IQR:.2f}")
    print(f"  Lower bound: {lower:.2f} | Upper bound: {upper:.2f}")
    print(f"  Outliers: {len(outliers):,} ({len(outliers)/len(df)*100:.1f}%)")
    print()

amount:
  Q1: 309.35 | Q3: 1175.62 | IQR: 866.27
  Lower bound: -990.05 | Upper bound: 2475.02
  Outliers: 1,828 (12.2%)

transaction_fee:
  Q1: 2.77 | Q3: 11.24 | IQR: 8.47
  Lower bound: -9.94 | Upper bound: 23.95
  Outliers: 1,871 (12.5%)

revenue_generated:
  Q1: 2.22 | Q3: 8.99 | IQR: 6.77
  Lower bound: -7.93 | Upper bound: 19.14
  Outliers: 1,871 (12.5%)



#### Decision: Retain Outliers
Outliers identified across amount, transaction_fee and 
revenue_generated represent legitimate high value 
transactions by Premium and Metal plan subscribers. 
Removing them would result in loss
 of valuable business insight. All outliers are retained.

## Feature Engineering
Extracting additional date and time features from 
transaction_date and transaction_time to support 
time-series analysis.



#### Extract date features from transaction_date

In [27]:
df['year'] = df['transaction_date'].dt.year
df['month'] = df['transaction_date'].dt.month
df['month_name'] = df['transaction_date'].dt.strftime('%B')
df['day_of_week'] = df['transaction_date'].dt.strftime('%A')
df['hour'] = pd.to_datetime(df['transaction_time'], 
format='%H:%M:%S').dt.hour
df

,transaction_id,user_id,transaction_date,transaction_time,transaction_type,transaction_category,amount,currency,transaction_status,payment_method,...,subscription_plan,fraud_flag,transaction_fee,revenue_generated,cashback_amount,year,month,month_name,day_of_week,hour
0,TXN007094,USR93203,2023-01-01,9:12:25,Transfer,Utilities,850.38,EUR,Successful,Card,...,Plus,0,8.50,6.80,4.25,2023,1,January,Sunday,9
1,TXN011155,USR17160,2023-01-01,13:45:55,Transfer,Shopping,4745.29,GBP,Successful,Bank Transfer,...,Standard,1,71.18,56.94,0.00,2023,1,January,Sunday,13
2,TXN002651,USR24022,2023-01-01,15:11:31,Purchase,Uncategorised,409.40,GBP,Successful,Apple Pay,...,Standard,0,6.14,4.91,0.00,2023,1,January,Sunday,15
3,TXN011642,USR85377,2023-01-01,21:01:06,Purchase,Food & Dining,643.73,EUR,Successful,Google Pay,...,Metal,0,1.29,1.03,9.66,2023,1,January,Sunday,21
4,TXN003802,USR80439,2023-01-01,22:24:18,Purchase,Shopping,143.16,GBP,Successful,Apple Pay,...,Metal,0,0.29,0.23,2.15,2023,1,January,Sunday,22
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14995,TXN001957,USR55745,2024-12-31,8:55:37,Top-up,Shopping,2021.10,GBP,Failed,Card,...,Plus,0,20.21,16.17,0.00,2024,12,December,Tuesday,8
14996,TXN007274,USR85454,2024-12-31,20:56:37,Purchase,Healthcare,418.49,EUR,Successful,Bank Transfer,...,Standard,0,6.28,5.02,0.00,2024,12,December,Tuesday,20
14997,TXN009711,USR55398,2024-12-31,22:09:14,Purchase,Shopping,630.43,GBP,Successful,Card,...,Standard,0,9.46,7.57,0.00,2024,12,December,Tuesday,22
14998,TXN014078,USR60014,2024-12-31,12:36:09,Transfer,Food & Dining,2544.17,USD,Successful,Card,...,Plus,0,25.44,20.35,12.72,2024,12,December,Tuesday,12


#### Final Verification
Confirming the dataset is fully clean and ready 
for analysis.

In [31]:
print("Missing values after cleaning:")
print(df.isnull().sum()[df.isnull().sum() > 0])
print("\nNo missing values ✅" if df.isnull().sum().sum() == 0 
else "⚠️ Missing values still exist")
print(f"\nFinal dataset shape: {df.shape}")
print("\nData types:")
print(df.dtypes)

Missing values after cleaning:
Series([], dtype: int64)

No missing values ✅

Final dataset shape: (15000, 23)

Data types:
transaction_id                  object
user_id                         object
transaction_date        datetime64[ns]
transaction_time                object
transaction_type                object
transaction_category            object
amount                         float64
currency                        object
transaction_status              object
payment_method                  object
age_group                       object
country                         object
city                            object
subscription_plan               object
fraud_flag                       int64
transaction_fee                float64
revenue_generated              float64
cashback_amount                float64
year                             int32
month                            int32
month_name                      object
day_of_week                     object
hour              

#### Save cleaned dataset

In [33]:
df.to_csv('../data/revolut_cleaned.csv', index=False)
print("Cleaned dataset saved successfully")
print(f"Shape: {df.shape}")

Cleaned dataset saved successfully
Shape: (15000, 23)
